sql: do add/drop column backfills via distsql

jordanlewis opened the pull request:
- Move `rowwriter` and `makeDefaultExprs` into `sqlbase` so they can be used from the `distsql` package.
- Replace the current implementation of backfilling an add/drop column schema change with one that uses distsql the way that the index backfiller works.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/14331)
<!-- Reviewable:end -->

jordanlewis committed:
sql: do add/drop column backfills via distsql

This commit replaces the add/drop column backfill methods in
`backfill.go` with a new `chunkBackfiller` called `columnBackfiller`
that does the backfill via distsql.

jordanlewis committed:
sql: move makeDefaultExprs to the sqlbase package

makeDefaultExprs is currently used by insert, update, upsert and
backfill, which all live in the sql package. It will soon by used by
distsqlrun, so it needed to be moved somewhere importable by both sql
and distsqlrun - it now lives in sqlbase.

jordanlewis committed:
sql: move rowwriter machinery to sqlbase

In preparation for being used by distsql.

jordanlewis committed:
distsqlrun: add fk table descs to backfill proto

jordanlewis committed:
distsqlrun: type assertion for indexbackfiller

jordanlewis commented:
This is out for early review. The implementation currently passes logic tests but does not pass `TestRaceWithBackfill`, so there is still some work to be done.

vivekmenezes commented:
Reviewed 5 of 5 files at r1, 13 of 13 files at r2, 1 of 1 files at r3, 2 of 2 files at r4.
Review status: 15 of 20 files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[pkg/sql/backfill.go, line 144 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/14331#-Kfszfjw4H-6wqpv6P9z:-Kfszfjw4H-6wqpv6PA-:blrrxui) ([raw file](https://github.com/cockroachdb/cockroach/blob/f621ea40277d49efc923a30bd679c7791591611e/pkg/sql/backfill.go#L144)):*
> ```Go
> 			switch t := m.Descriptor_.(type) {
> 			case *sqlbase.DescriptorMutation_Column:
> 				addedColumnDescs = append(addedColumnDescs, *t.Column)
> ```

we don't need addedColumnDescs anymore

---

*[pkg/sql/backfill.go, line 164 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/14331#-KfszypdvsLR4KLxjuzL:-KfszypdvsLR4KLxjuzM:b-iscs5c) ([raw file](https://github.com/cockroachdb/cockroach/blob/f621ea40277d49efc923a30bd679c7791591611e/pkg/sql/backfill.go#L164)):*
> ```Go
> 			switch t := m.Descriptor_.(type) {
> 			case *sqlbase.DescriptorMutation_Column:
> 				droppedColumnDescs = append(droppedColumnDescs, *t.Column)
> ```

needColumnBackfill = true 

we use the same backfill mechanism to delete a column

get rid of droppedColumnDescs

---

*[pkg/sql/distsql_physical_planner.go, line 622 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/14331#-Kft01Fi4Q-I5WWaQz0A:-Kft01Fi4Q-I5WWaQz0B:b-7t17q1) ([raw file](https://github.com/cockroachdb/cockroach/blob/f621ea40277d49efc923a30bd679c7791591611e/pkg/sql/distsql_physical_planner.go#L622)):*
> ```Go
> 			Duration:    duration,
> 			ChunkSize:   chunkSize,
> 			OtherTables: fkDescs,
> ```

might as well create a generic distsqlrun.BackfillerSpec and only use the switch to set Type

---

*[pkg/sql/distsqlrun/columnbackfiller.go, line 36 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/14331#-Kft0T61lLlOuTVxgk8p:-Kft0T61lLlOuTVxgk8q:b-tlqsw2) ([raw file](https://github.com/cockroachdb/cockroach/blob/f621ea40277d49efc923a30bd679c7791591611e/pkg/sql/distsqlrun/columnbackfiller.go#L36)):*
> ```Go
> 	added        []sqlbase.ColumnDescriptor
> 	dropped      []sqlbase.ColumnDescriptor
> 	updateCols   []sqlbase.ColumnDescriptor
> ```

updateCols?

---

*[pkg/sql/distsqlrun/columnbackfiller.go, line 41 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/14331#-Kft0blclGNYp7qd1Wnt:-Kft0blclGNYp7qd1Wnu:b-8vp86x) ([raw file](https://github.com/cockroachdb/cockroach/blob/f621ea40277d49efc923a30bd679c7791591611e/pkg/sql/distsqlrun/columnbackfiller.go#L41)):*
> ```Go
> 	nonNullViolationColumnName string
> 
> 	// colIdxMap maps ColumnIDs to indices into desc.Columns and desc.Mutations.
> ```

and desc.Mutations that hold column descriptors.

---

*[pkg/sql/distsqlrun/columnbackfiller.go, line 89 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/14331#-Kft1FHi9omI6l5Du9L4:-Kft1FHj491f1DwPYC9z:bz1s8bk) ([raw file](https://github.com/cockroachdb/cockroach/blob/f621ea40277d49efc923a30bd679c7791591611e/pkg/sql/distsqlrun/columnbackfiller.go#L89)):*
> ```Go
> 		for _, m := range desc.Mutations {
> 			if ColumnMutationFilter(m) {
> 				switch m.Direction {
> ```

I think you meant to also append the mutation column to cols

---

*[pkg/sql/distsqlrun/columnbackfiller.go, line 134 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/14331#-Kft2-i7Ve0e6wKf0E7m:-Kft2-i7Ve0e6wKf0E7n:b-ki5tyi) ([raw file](https://github.com/cockroachdb/cockroach/blob/f621ea40277d49efc923a30bd679c7791591611e/pkg/sql/distsqlrun/columnbackfiller.go#L134)):*
> ```Go
> 	}
> 
> 	// We need all the columns.
> ```

I think we do not need all the columns. Perhaps add a TODO to use a smaller set like the set of columns in the primary key. We basically want to create an insert event per row.

---

*[pkg/sql/distsqlrun/columnbackfiller.go, line 164 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/14331#-Kft2SQrGjDxwHr8Yl6v:-Kft2SQrGjDxwHr8Yl6w:b-t1vfdj) ([raw file](https://github.com/cockroachdb/cockroach/blob/f621ea40277d49efc923a30bd679c7791591611e/pkg/sql/distsqlrun/columnbackfiller.go#L164)):*
> ```Go
> 		}
> 
> 		// TODO(vivek): is this necessary?
> ```

we don't need it!

---

*[pkg/sql/distsqlrun/processors.proto, line 360 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/14331#-Kfsy_mlQaBRZIsKx5O0:-Kfsy_mlQaBRZIsKx5O1:b-q5ce87) ([raw file](https://github.com/cockroachdb/cockroach/blob/e2c298013a0df7ee49aef444ada5796d5a09e32b/pkg/sql/distsqlrun/processors.proto#L360)):*
> ```protobuf
>   // backfiller to do its job, such as the descriptors for tables with fk
>   // relationships to the table being modified.
>   repeated sqlbase.TableDescriptor otherTables = 6 [(gogoproto.nullable) = false];
> ```

other_tables

---

*[pkg/sql/sqlbase/rowwriter.go, line 761 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14331#-Kfsxr6o-mJCZr5LPG5z:-Kfsxr6p1jYGxYrlfWX_:b-xwsd4r) ([raw file](https://github.com/cockroachdb/cockroach/blob/6f9a7c81dab26c46b366bec4c811903482c22fe1/pkg/sql/sqlbase/rowwriter.go#L761)):*
> ```Go
> 	Helper               rowHelper
> 	FetchCols            []ColumnDescriptor
> 	FetchColIDtoRowIndex map[ColumnID]int
> ```

Did you really need to export these as part of this PR?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14331)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
I took a first pass of this but would like to review runChunk() again.

I actually forget why we care about FKs for column backfill and will look into that tomorrow.

---

Review status: 15 of 20 files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14331#-:-Kft2uWTQGZEPZUmRJPK:bt9y2al)*
<!-- Sent from Reviewable.io -->

jordanlewis commented:
Review status: 13 of 20 files reviewed at latest revision, 10 unresolved discussions.

---

*[pkg/sql/backfill.go, line 144 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/14331#-Kfszfjw4H-6wqpv6P9z:-Kfvn9rKNdNwyWi8dKPy:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f621ea40277d49efc923a30bd679c7791591611e/pkg/sql/backfill.go#L144)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

we don't need addedColumnDescs anymore
</blockquote></details>

Done.

---

*[pkg/sql/backfill.go, line 164 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/14331#-KfszypdvsLR4KLxjuzL:-KfvnACcKJnUbHP50qlH:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f621ea40277d49efc923a30bd679c7791591611e/pkg/sql/backfill.go#L164)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

needColumnBackfill = true 

we use the same backfill mechanism to delete a column

get rid of droppedColumnDescs
</blockquote></details>

Done.

---

*[pkg/sql/distsqlrun/columnbackfiller.go, line 36 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/14331#-Kft0T61lLlOuTVxgk8p:-KfvnKY0RGfUXA2VDqwU:b-f8sf1p) ([raw file](https://github.com/cockroachdb/cockroach/blob/f621ea40277d49efc923a30bd679c7791591611e/pkg/sql/distsqlrun/columnbackfiller.go#L36)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

updateCols?
</blockquote></details>

Added a comment.

---

*[pkg/sql/distsqlrun/columnbackfiller.go, line 41 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/14331#-Kft0blclGNYp7qd1Wnt:-KfvnRO2QEL64JM-5rF_:bhiocvn) ([raw file](https://github.com/cockroachdb/cockroach/blob/f621ea40277d49efc923a30bd679c7791591611e/pkg/sql/distsqlrun/columnbackfiller.go#L41)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

and desc.Mutations that hold column descriptors.
</blockquote></details>

I don't understand what you mean?

---

*[pkg/sql/distsqlrun/columnbackfiller.go, line 89 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/14331#-Kft1FHi9omI6l5Du9L4:-KfvsDTQJwV-UhqN5_gd:b-ttb257) ([raw file](https://github.com/cockroachdb/cockroach/blob/f621ea40277d49efc923a30bd679c7791591611e/pkg/sql/distsqlrun/columnbackfiller.go#L89)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

I think you meant to also append the mutation column to cols
</blockquote></details>

Yes, thanks! This fixed the test issue :)

---

*[pkg/sql/distsqlrun/columnbackfiller.go, line 164 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/14331#-Kft2SQrGjDxwHr8Yl6v:-KfvtnQl6ciTgKHLFAGX:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f621ea40277d49efc923a30bd679c7791591611e/pkg/sql/distsqlrun/columnbackfiller.go#L164)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

we don't need it!
</blockquote></details>

Done.

---

*[pkg/sql/distsqlrun/processors.proto, line 360 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/14331#-Kfsy_mlQaBRZIsKx5O0:-KfvtmuJC1Jysqryp5cH:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e2c298013a0df7ee49aef444ada5796d5a09e32b/pkg/sql/distsqlrun/processors.proto#L360)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

other_tables
</blockquote></details>

Done.

---

*[pkg/sql/sqlbase/rowwriter.go, line 761 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14331#-Kfsxr6o-mJCZr5LPG5z:-KfvsJYYs5nz-BeAZKLK:b-1k01l2) ([raw file](https://github.com/cockroachdb/cockroach/blob/6f9a7c81dab26c46b366bec4c811903482c22fe1/pkg/sql/sqlbase/rowwriter.go#L761)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

Did you really need to export these as part of this PR?
</blockquote></details>

Unfortunately, yes. All of these structs needed to move to sqlbase, and in order to do that, I needed make a bunch of these fields public.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14331)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Review status: 13 of 20 files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[pkg/sql/distsqlrun/columnbackfiller.go, line 41 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/14331#-Kft0blclGNYp7qd1Wnt:-KfwC7RoRDikgQ4puWyg:btze2tj) ([raw file](https://github.com/cockroachdb/cockroach/blob/f621ea40277d49efc923a30bd679c7791591611e/pkg/sql/distsqlrun/columnbackfiller.go#L41)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

I don't understand what you mean?
</blockquote></details>

the mutations can be either column or index mutations. But actually it's better the way it is because it is an index into Mutations[] independent of it hold. SO leave this as is.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14331)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented on pkg/sql/distsqlrun/processors.proto:
```diff
@@ -353,6 +353,11 @@ message BackfillerSpec {
+  // other_tables contains any other (leased) table descriptors necessary for the
+  // backfiller to do its job, such as the descriptors for tables with fk
+  // relationships to the table being modified.
+  repeated sqlbase.TableDescriptor other_tables = 6 [(gogoproto.nullable) = false];
```
As discussed foreign key constraints can be checked before and after column backfill and so need not be passed to each processor.

tamird commented:
3rd commit message lacks a verb: "add"?

---

Reviewed 5 of 5 files at r1, 13 of 13 files at r2, 1 of 1 files at r3, 7 of 7 files at r6, 5 of 5 files at r7.
Review status: all files reviewed at latest revision, 17 unresolved discussions, some commit checks failed.

---

*[pkg/sql/backfill.go, line 407 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/14331#-KfwbZ3iSL-lkBGNmIV-:-KfwbZ3iSL-lkBGNmIV0:b-z1m0rp) ([raw file](https://github.com/cockroachdb/cockroach/blob/19df37820ddf6be4f8f3771057d3299a0f77be97/pkg/sql/backfill.go#L407)):*
> ```Go
> 				return err
> 			}
> 			var fkTableDescs []sqlbase.TableDescriptor
> ```

this should be otherTableDescs, right? this might include other things in the future.

---

*[pkg/sql/distsql_physical_planner.go, line 604 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/14331#-KfwbyxfcHSxIM45N8D8:-KfwbyxfcHSxIM45N8D9:bgihvbj) ([raw file](https://github.com/cockroachdb/cockroach/blob/19df37820ddf6be4f8f3771057d3299a0f77be97/pkg/sql/distsql_physical_planner.go#L604)):*
> ```Go
> 	duration time.Duration,
> 	chunkSize int64,
> 	fkDescs []sqlbase.TableDescriptor,
> ```

otherTables?

---

*[pkg/sql/distsql_physical_planner.go, line 640 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/14331#-Kfwbwd1Re09Ekqv3AT-:-Kfwbwd21su9G9tvqYRZ:bgihvbj) ([raw file](https://github.com/cockroachdb/cockroach/blob/19df37820ddf6be4f8f3771057d3299a0f77be97/pkg/sql/distsql_physical_planner.go#L640)):*
> ```Go
> 	chunkSize int64,
> 	spans []roachpb.Span,
> 	fkDescs []sqlbase.TableDescriptor,
> ```

otherTables?

---

*[pkg/sql/distsqlrun/columnbackfiller.go, line 1 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/14331#-KfwPGZXPncdzuONe2Ox:-KfwPGZXPncdzuONe2Oy:bb6ht2h) ([raw file](https://github.com/cockroachdb/cockroach/blob/19df37820ddf6be4f8f3771057d3299a0f77be97/pkg/sql/distsqlrun/columnbackfiller.go#L1)):*
> ```Go
> // Copyright 2016 The Cockroach Authors.
> ```

2017

---

*[pkg/sql/distsqlrun/processors.proto, line 357 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/14331#-KfwOFdmDN27GMf80Q9_:-KfwOFdmDN27GMf80Q9a:b306u4k) ([raw file](https://github.com/cockroachdb/cockroach/blob/3217ec1ff5faf0e06254cba3796aa1f49e91ea20/pkg/sql/distsqlrun/processors.proto#L357)):*
> ```protobuf
>   optional int64 chunk_size = 5 [(gogoproto.nullable) = false];
> 
>   // other_tables contains any other (leased) table descriptors necessary for the
> ```

Looks like the local convention avoids naming the field in its comment.

---

*[pkg/sql/sqlbase/fk.go, line 32 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14331#-KfwNlFfL7qlRvrLlQ_y:-KfwNlFgeqEAvn-HlCHq:b-1kpdsm) ([raw file](https://github.com/cockroachdb/cockroach/blob/6f9a7c81dab26c46b366bec4c811903482c22fe1/pkg/sql/sqlbase/fk.go#L32)):*
> ```Go
> type TableLookupsByID map[ID]TableLookup
> 
> // TableLookup is the value type of TableLookupsByID: An optional table
> ```

s/An/an/

but this comment would be better on each field rather than here.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14331#-:-KfwO5bcsiQif693egBO:b-ob5is)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 18 unresolved discussions, some commit checks failed.

---

*[pkg/sql/distsqlrun/columnbackfiller.go, line 64 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/14331#-Kfwc_N_GAClZVzv8Le1:-Kfwc_N_GAClZVzv8Le2:b-a6gbu) ([raw file](https://github.com/cockroachdb/cockroach/blob/19df37820ddf6be4f8f3771057d3299a0f77be97/pkg/sql/distsqlrun/columnbackfiller.go#L64)):*
> ```Go
> 			flowCtx: flowCtx,
> 			output:  output,
> 			spec:    *spec,
> ```

should this deref in the caller? that's where the nil check is.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14331)*
<!-- Sent from Reviewable.io -->

jordanlewis committed:
distsqlrun: pass index backfiller spec by value

Previously, the backfiller spec was passed to the index backfiller by
reference, despite it getting dereferenced immediately afterwards and
despite the caller being the one performing the nil check. Now,
dereference the spec right after the nil check is done.

jordanlewis commented:
Review status: 12 of 20 files reviewed at latest revision, 18 unresolved discussions.

---

*[pkg/sql/backfill.go, line 407 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/14331#-KfwbZ3iSL-lkBGNmIV-:-KfxZ3VR4Ymb0iygJLmQ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/19df37820ddf6be4f8f3771057d3299a0f77be97/pkg/sql/backfill.go#L407)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

this should be otherTableDescs, right? this might include other things in the future.
</blockquote></details>

Done.

---

*[pkg/sql/distsql_physical_planner.go, line 622 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/14331#-Kft01Fi4Q-I5WWaQz0A:-KfxZa21PuJsMb2bSqJu:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f621ea40277d49efc923a30bd679c7791591611e/pkg/sql/distsql_physical_planner.go#L622)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

might as well create a generic distsqlrun.BackfillerSpec and only use the switch to set Type
</blockquote></details>

Done.

---

*[pkg/sql/distsql_physical_planner.go, line 604 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/14331#-KfwbyxfcHSxIM45N8D8:-KfxZOQQxu4DuAuy_lDJ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/19df37820ddf6be4f8f3771057d3299a0f77be97/pkg/sql/distsql_physical_planner.go#L604)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

otherTables?
</blockquote></details>

Done.

---

*[pkg/sql/distsql_physical_planner.go, line 640 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/14331#-Kfwbwd1Re09Ekqv3AT-:-KfxZelCFdTAvhrOZfS5:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/19df37820ddf6be4f8f3771057d3299a0f77be97/pkg/sql/distsql_physical_planner.go#L640)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

otherTables?
</blockquote></details>

Done.

---

*[pkg/sql/distsqlrun/columnbackfiller.go, line 1 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/14331#-KfwPGZXPncdzuONe2Ox:-KfxZjGKCDv0jSFRkN7e:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/19df37820ddf6be4f8f3771057d3299a0f77be97/pkg/sql/distsqlrun/columnbackfiller.go#L1)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

2017
</blockquote></details>

Done.

---

*[pkg/sql/distsqlrun/columnbackfiller.go, line 64 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/14331#-Kfwc_N_GAClZVzv8Le1:-Kfx_7q54iQU5RYmteKA:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/19df37820ddf6be4f8f3771057d3299a0f77be97/pkg/sql/distsqlrun/columnbackfiller.go#L64)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

should this deref in the caller? that's where the nil check is.
</blockquote></details>

Done.

---

*[pkg/sql/distsqlrun/processors.proto, line 357 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/14331#-KfwOFdmDN27GMf80Q9_:-Kfx_JFNCyXFhsMm2alL:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3217ec1ff5faf0e06254cba3796aa1f49e91ea20/pkg/sql/distsqlrun/processors.proto#L357)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

Looks like the local convention avoids naming the field in its comment.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14331)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 13 files at r2, 1 of 8 files at r8, 2 of 2 files at r9, 5 of 5 files at r10.
Review status: all files reviewed at latest revision, 12 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14331)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Exciting stuff!

---

Review status: all files reviewed at latest revision, 12 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14331#-:-Kg-hOk7F13ypfZd3IIX:bujbiap)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Do let me know when I should take a look

On Fri, Mar 24, 2017, 10:27 AM RaduBerinde <notifications@github.com> wrote:

> Exciting stuff!
> ------------------------------
>
> Review status: all files reviewed at latest revision, 12 unresolved
> discussions, some commit checks failed.
> ------------------------------
>
> *Comments from Reviewable
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/14331#-:-Kg-hOk7F13ypfZd3IIX:bujbiap>*
>
> —
> You are receiving this because your review was requested.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/14331#issuecomment-289037199>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ALOpBKVAinf2Ui5z3JjXuWfUQJbOPAQKks5ro9LAgaJpZM4Ml8ER>
> .
>

jordanlewis commented:
PTAL!

---

Review status: 2 of 21 files reviewed at latest revision, 12 unresolved discussions, some commit checks pending.

---

*[pkg/sql/distsqlrun/columnbackfiller.go, line 134 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/14331#-Kft2-i7Ve0e6wKf0E7m:-KgQRKdM8N95ehH2Dukd:b-6x9yrt) ([raw file](https://github.com/cockroachdb/cockroach/blob/f621ea40277d49efc923a30bd679c7791591611e/pkg/sql/distsqlrun/columnbackfiller.go#L134)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

I think we do not need all the columns. Perhaps add a TODO to use a smaller set like the set of columns in the primary key. We basically want to create an insert event per row.
</blockquote></details>

There's already a todo for this in this file about tightening the bounds on the requestedCols down below in the rowUpdater. That todo encapsulates this request.

---

*[pkg/sql/distsqlrun/processors.proto, line 360 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/14331#-KfsGyHxbCFqd_NcTeLe-r7-360:-KgQLZgFkPwSeX5byho5:b8atay) ([raw file](https://github.com/cockroachdb/cockroach/blob/19df37820ddf6be4f8f3771057d3299a0f77be97/pkg/sql/distsqlrun/processors.proto#L360)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

As discussed foreign key constraints can be checked before and after column backfill and so need not be passed to each processor.
</blockquote></details>

I'll do this in a subsequent PR as discussed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14331#-:-KgQYy_vDH60f6NN-X-9:b-83330z)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 2 of 21 files reviewed at latest revision, 17 unresolved discussions, all commit checks successful.

---

*[pkg/sql/backfill.go, line 144 at r16](https://reviewable.io:443/reviews/cockroachdb/cockroach/14331#-KgQno-6cNnIU6PCU-iG:-KgQno-6cNnIU6PCU-iH:b9mjykf) ([raw file](https://github.com/cockroachdb/cockroach/blob/c7df0b8f9a02aa8fddc67db12859bb6f60534523/pkg/sql/backfill.go#L144)):*
> ```Go
> 				desc := m.GetColumn()
> 				if desc.DefaultExpr != nil || !desc.Nullable {
> 					needColumnBackfill = true
> ```

created this issue: https://github.com/cockroachdb/cockroach/issues/14455 to clean this up

---

*[pkg/sql/backfill.go, line 458 at r16](https://reviewable.io:443/reviews/cockroachdb/cockroach/14331#-KgRxhyMwOOHF637VVhE:-KgRxhyMwOOHF637VVhF:b3w5rxb) ([raw file](https://github.com/cockroachdb/cockroach/blob/c7df0b8f9a02aa8fddc67db12859bb6f60534523/pkg/sql/backfill.go#L458)):*
> ```Go
> 	version sqlbase.DescriptorVersion,
> ) error {
> 	return sc.distBackfill(ctx, lease, version, columnBackfill, columnTruncateAndBackfillChunkSize, distsqlrun.ColumnMutationFilter)
> ```

can you use getChunkSize(columnTruncateAndBackfillChunkSize)

please also fix the backfillIndexes case to use

getChunkSize(indexBackfillChunkSize) in a separate commit. Thanks!

---

*[pkg/sql/distsql_physical_planner.go, line 615 at r16](https://reviewable.io:443/reviews/cockroachdb/cockroach/14331#-KgRyE2kCzQFPqI5uByf:-KgRyE2kCzQFPqI5uByg:b-9v2f2i) ([raw file](https://github.com/cockroachdb/cockroach/blob/c7df0b8f9a02aa8fddc67db12859bb6f60534523/pkg/sql/distsql_physical_planner.go#L615)):*
> ```Go
> 	case indexBackfill:
> 		ret.Type = distsqlrun.BackfillerSpec_Index
> 		return ret, nil
> ```

call return ret, nil at the bottom

---

*[pkg/sql/distsqlrun/columnbackfiller.go, line 82 at r16](https://reviewable.io:443/reviews/cockroachdb/cockroach/14331#-KgRzJNBLq_8OusvBJXe:-KgRzJNBLq_8OusvBJXf:b-k9oo1l) ([raw file](https://github.com/cockroachdb/cockroach/blob/c7df0b8f9a02aa8fddc67db12859bb6f60534523/pkg/sql/distsqlrun/columnbackfiller.go#L82)):*
> ```Go
> 	// If that's the case, and we end up reading a non-zero amount of data,
> 	// we need a throw an error since the old columns will already violate the
> 	// not null constraint.
> ```

TODO(jordan): fix #14455

---

*[pkg/sql/distsqlrun/columnbackfiller.go, line 132 at r16](https://reviewable.io:443/reviews/cockroachdb/cockroach/14331#-KgS-XuVg1rqwj9_8R86:-KgS-XuVg1rqwj9_8R87:b-sighbo) ([raw file](https://github.com/cockroachdb/cockroach/blob/c7df0b8f9a02aa8fddc67db12859bb6f60534523/pkg/sql/distsqlrun/columnbackfiller.go#L132)):*
> ```Go
> 	}
> 
> 	// We need all the columns.
> ```

Added a TODO that we need to only read one of the columns forming the primary key. It's better not to make that change in this PR.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14331#-:-KgS16UsmR6mM1Q1hBVe:bnfp4nl)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
#12424

jordanlewis commented:
Review status: 2 of 21 files reviewed at latest revision, 17 unresolved discussions, all commit checks successful.

---

*[pkg/sql/backfill.go, line 458 at r16](https://reviewable.io:443/reviews/cockroachdb/cockroach/14331#-KgRxhyMwOOHF637VVhE:-KgUs4iFy1A7zKyl2MeN:b859dhc) ([raw file](https://github.com/cockroachdb/cockroach/blob/c7df0b8f9a02aa8fddc67db12859bb6f60534523/pkg/sql/backfill.go#L458)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

can you use getChunkSize(columnTruncateAndBackfillChunkSize)

please also fix the backfillIndexes case to use

getChunkSize(indexBackfillChunkSize) in a separate commit. Thanks!
</blockquote></details>

I'll fix these both in a follow on.

---

*[pkg/sql/distsql_physical_planner.go, line 615 at r16](https://reviewable.io:443/reviews/cockroachdb/cockroach/14331#-KgRyE2kCzQFPqI5uByf:-KgUsPI_yiQo4fSyQT_G:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c7df0b8f9a02aa8fddc67db12859bb6f60534523/pkg/sql/distsql_physical_planner.go#L615)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

call return ret, nil at the bottom
</blockquote></details>

Done.

---

*[pkg/sql/distsqlrun/columnbackfiller.go, line 82 at r16](https://reviewable.io:443/reviews/cockroachdb/cockroach/14331#-KgRzJNBLq_8OusvBJXe:-KgUsgJQLd_I-ZPmWhCk:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c7df0b8f9a02aa8fddc67db12859bb6f60534523/pkg/sql/distsqlrun/columnbackfiller.go#L82)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

TODO(jordan): fix #14455
</blockquote></details>

Done.

---

*[pkg/sql/distsqlrun/columnbackfiller.go, line 132 at r16](https://reviewable.io:443/reviews/cockroachdb/cockroach/14331#-KgS-XuVg1rqwj9_8R86:-KgUshXbFVY-8RKYUSTQ:b-iu52of) ([raw file](https://github.com/cockroachdb/cockroach/blob/c7df0b8f9a02aa8fddc67db12859bb6f60534523/pkg/sql/distsqlrun/columnbackfiller.go#L132)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

Added a TODO that we need to only read one of the columns forming the primary key. It's better not to make that change in this PR.
</blockquote></details>

Hmm, I'm not sure that this is true. What if we're adding a new column that's part of the same column family as the rest of the columns? Granted the correct bounds for this is more subtle, but agreed that I'll do this in a separate PR.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14331)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Reviewed 2 of 13 files at r2, 2 of 13 files at r12, 19 of 21 files at r17.
Review status: 2 of 21 files reviewed at latest revision, 17 unresolved discussions, all commit checks successful.

---

*[pkg/sql/distsqlrun/columnbackfiller.go, line 132 at r16](https://reviewable.io:443/reviews/cockroachdb/cockroach/14331#-KgS-XuVg1rqwj9_8R86:-KgV6OMlG6FOi_str-Oa:b32wtd) ([raw file](https://github.com/cockroachdb/cockroach/blob/c7df0b8f9a02aa8fddc67db12859bb6f60534523/pkg/sql/distsqlrun/columnbackfiller.go#L132)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Hmm, I'm not sure that this is true. What if we're adding a new column that's part of the same column family as the rest of the columns? Granted the correct bounds for this is more subtle, but agreed that I'll do this in a separate PR.
</blockquote></details>

ok

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14331)*
<!-- Sent from Reviewable.io -->

jordanlewis commented:
TFTRs!

---

Review status: 2 of 21 files reviewed at latest revision, 16 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14331#-:-KgVRIQ6rX1xna33sJoO:bz5mair)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
woohoo

On Thu, Mar 30, 2017 at 2:19 PM Jordan Lewis <notifications@github.com>
wrote:

> Merged #14331 <https://github.com/cockroachdb/cockroach/pull/14331>.
>
> —
> You are receiving this because your review was requested.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/14331#event-1022830409>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ALOpBEx61_hA4UNw28IM4YrGVMf2TFRcks5rq_IzgaJpZM4Ml8ER>
> .
>
